numba_rvsdg/core/datastructures/ast_transforms.py

```diff
@@ -0,0 +1,603 @@
+        to_visit, reachable, unreachable = set("0"), set(), set()
+        # Visit all reachable blocks.
+        while to_visit:
+            block = to_visit.pop()
+            if block not in reachable:
+                # Add block to reachable set.
+                reachable.add(block)
+                # Update to_visit with jump targets of the block.
+                to_visit.update(self[block].jump_targets)
+        # Remove unreachable blocks.
+        for block in list(self.keys()):
+            if block not in reachable:
+                unreachable.add(self.pop(block))
+        self.unreachable = unreachable
+        return unreachable
+
+    def prune_empty(self) -> set[WritableASTBlock]:
+        """Prune empty blocks from the CFG."""
+        empty = set()
+        for name, block in list(self.items()):
+            if not block.instructions:
+                empty.add(self.pop(name))
+                # Empty blocks can only have a single jump target.
+                it = block.jump_targets[0]
+                # Iterate over the blocks looking for blocks that point to the
+                # removed block. Then rewire the jump_targets accordingly.
+                for b in list(self.values()):
+                    if len(b.jump_targets) == 0:
+                        continue
+                    elif len(b.jump_targets) == 1:
+                        if b.jump_targets[0] == name:
+                            b.jump_targets[0] = it
+                    elif len(b.jump_targets) == 2:
+                        if b.jump_targets[0] == name:
+                            b.jump_targets[0] = it
+                        elif b.jump_targets[1] == name:
+                            b.jump_targets[1] = it
+        self.empty = empty
+        return empty
+
+
+@dataclass(frozen=True)
+class LoopIndices:
+    """Structure to hold the head and exit block indices of a loop."""
+
+    head: int
+    exit: int
+
+
+class AST2SCFGTransformer:
+    """AST2SCFGTransformer
+
+    The AST2SCFGTransformer class is responsible for transforming code in the
+    form of a Python Abstract Syntax Tree (AST) into CFG/SCFG.
+
+    """
+
+    def __init__(self, code: Callable[..., Any], prune: bool = True) -> None:
+        # Prune empty and unreachable blocks from the CFG.
+        self.prune: bool = prune
+        # Save the code for transformation.
+        self.code: Callable[..., Any] = code
+        # Monotonically increasing block index, 0 is reserved for genesis.
+        self.block_index: int = 1
+        # Dict mapping block indices as strings to WritableASTBlocks.
+        # (This is the data structure to hold the CFG.)
+        self.blocks: ASTCFG = ASTCFG()
+        # Initialize first (genesis) block, assume it's named zero.
+        # (This also initializes the self.current_block attribute.)
+        self.add_block(0)
+        # Stack for header and exiting block of current loop.
+        self.loop_stack: list[LoopIndices] = []
```

Comment:
Type annotation for instance variables should be in class scope. See https://peps.python.org/pep-0526/#class-and-instance-variable-annotations

Reply:
done in 8a119a4bc835362dc149a6bc4d85e28d972ec3cd